### ISSUE TITLE
Add an IBodyProducer which produces bytes from a file-like object

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4017]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#4017|
|Type|enhancement|
|Created|2009-09-11 13:41:13Z|
|Branch|https://github.com/twisted/twisted/tree/filebodyproducer-4017-2|

[#10610](https://github.com/twisted/twisted/issues/10610) introduces an HTTP client API along with a new interface, `IBodyProducer`, for providing request bodies in a streaming manner.  This interface isn't quite the same as `IProducer`, so existing producers don't quite work with it.

For convenience, we should provide an implementation of this interface based on file-like objects (along the lines of `twisted.protocols.basic.FileSender`).  This will be useful both for the obvious reason (sending files as request bodies), as well as as an example of how to implement these, and will also be usable with `StringIO` instances to send data constructed in memory (eg simple form posts).

<details><summary>Searchable metadata</summary>

```
trac-id__4017 4017
type__enhancement enhancement
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_filebodyproducer_4017_2 branches/filebodyproducer-4017-2
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__web web
keywords__httpclient_IBodyProducer httpclient IBodyProducer
time__1252676473000000 1252676473000000
changetime__1308876192000000 1308876192000000
version__None None
owner__exarkun exarkun

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="djfroofy's avatar" src="https://avatars.githubusercontent.com/u/89760?s=50" width="50" height="50">](https://github.com/djfroofy)|@djfroofy set owner to @djfroofy<br>@djfroofy set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

(In [32067]) Branching to 'filebodyproducer-4017'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

I took a crack at this because I thought it would be quick and easy.  Oh well, I've certainly done worse than none out of two before.

Please find `FileBodyProducer`, tests, docs, and an example in the linked branch.  Here are the [build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/filebodyproducer-4017).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

One possible change that should be made to this code is that the producer should take responsibility for closing the file after it finishes reading from it or if `stopProducing` is called.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_9"></a>|@therve set owner to @exarkun|
|-|-|

 * Instead of keeping a reference to the the cooperator object, you may only keep one to the cooperate method. It would prevent the need of that comment about the task module.
 * I don't have a really strong feeling for closing the file or not, but whatever you choose, it should be mentioned in the documentation (both the docstring and the howto).

Please merge!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

(In [32154]) Address review feedback, plus

  * add docstrings to some test helpers
  * use the bound cooperate method instead of a Cooperator instance/cooperator module
  * close the input file when the FileBodyProducer is done with it (and update docs)


refs [#4017](https://github.com/twisted/twisted/issues/4017)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

(In [32155]) Branching to 'filebodyproducer-4017-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun set status to `closed`|
|-|-|

(In [32157]) Merge filebodyproducer-4017-2

Author: exarkun
Reviewer: therve
Fixes: [#4017](https://github.com/twisted/twisted/issues/4017)

Add `twisted.web.client.FileBodyProducer` for use with `Agent` to specify request
bodies based on file-like objects.
